keps/sig-api-machinery/4420-retry-generate-name/README.md

```diff
@@ -0,0 +1,541 @@
+increases significantly when this feature is enabled, it may indicate
+an issue with this feature.
+
+We also plan to add a `generate_name_retries: {count}` label to this metric
+to make it easy to inspect the latency of create requests where retries
+occurred.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+N/A
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Presence of the `generate_name_retries: {count}` label on the `apiserver_request_duration_seconds` metric
+with a count greater than 1 would indicate this feature has been used.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Other
+  - Details: The lack of name conflict errors when using generateName is the only user visible behavior, but it
+    is non-deterministic.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+Create latency should remain unchanged because retries should be very rare under normal
+operating conditions. I.e. `apiserver_request_duration_seconds` should not change.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+<!--
+Pick one more of these and delete the rest.
+-->
+
+- [x] Metrics
+  - Metric name: `apiserver_request_duration_seconds` for creates
+
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+No
+
+### Dependencies
+
+###### Does this feature depend on any specific services running in the cluster?
+
+No
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+
+Yes, if the resource being created with generateName has validating webhooks registered,
+the validating webhooks will be called for each retry.
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+No
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+This can happen.
+
+When a retry does occur, if the resource being created with generateName has
+validating webhooks registered, the validating webhooks will be called for each
+retry. For slow webhooks this could result in a significant increase to request
+latency for the create requests using generateName.
```

Comment:
(nit)
```suggestion
When a retry does occur, if the resource being created with `generateName` has
validating webhooks registered, the validating webhooks will be called for each
retry. For slow webhooks this could result in a significant increase to request
latency for the create requests using `generateName`.
```

BTW same thing for in-process evaluation of ValidatingAdmissionPolicies.

---

We might want to make the retry count load adaptive. Maybe. One way to do that is to have the API server watch (as in local inotify) for a dynamic configuration snippet that controls this kind of setting, and then use a mechanism (ConfigMap? special systemd unit? whatever) to write an appropriate value to that dynamic overrides file.

For alpha, we absolutely don't need that mechanism.

Reply:
Yeah, I've been thinking about this a bit too.  We could either offer more controller, or we could optimize it further.  I'll capture all of this as potential future work (either as post alpha or post ga)